sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,305 @@
+                Err(_) => error::BadIpAddressVecSnafu.fail(),
+            }
+        }
+        16 => {
+            // We checked that its exactly 16 u8 long, forcing std::result::Result explicity to raise a nice error
+            let octet_result: std::result::Result<[u8; 16], _> = address_vec.try_into();
+            match octet_result {
+                Ok(octets) => Ok(IpAddr::from(octets)),
+                Err(_) => error::BadIpAddressVecSnafu.fail(),
+            }
+        }
+        _ => error::BadIpAddressSnafu {
+            input: address_vec,
+            msg: "invalid length, must be 4 or 16 octets".to_string(),
+        }
+        .fail(),
+    }
+}
+
+fn ipaddr_from_vec_de<'de, D>(deserializer: D) -> std::result::Result<IpAddr, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let address_vec: Vec<u8> = Deserialize::deserialize(deserializer)?;
+    ipaddr_from_vec(address_vec).map_err(D::Error::custom)
+}
+
+fn from_networkctl_addresses<'de, D>(deserializer: D) -> std::result::Result<Vec<IpAddr>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    #[derive(Debug, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    struct NetworkctlAddress {
+        address: Vec<u8>,
+    }
+    let addresses: Vec<NetworkctlAddress> = Deserialize::deserialize(deserializer)?;
+    let mut addrs = Vec::new();
+    for addr in addresses.iter() {
+        addrs.push(ipaddr_from_vec(addr.address.clone()).map_err(D::Error::custom)?);
+    }
+    Ok(addrs)
+}
+
+impl NetworkDInterfaceStatus {
+    pub(crate) fn new(link: String) -> Result<Self> {
+        let systemd_networkctl_result = Command::new(NETWORKCTL)
+            .arg("status")
+            .arg("--json=pretty")
+            .arg(link)
+            .output()
+            .context(error::NetworkctlExecutionSnafu)?;
+        ensure!(
+            systemd_networkctl_result.status.success(),
+            error::FailedNetworkctlSnafu {
+                stderr: String::from_utf8_lossy(&systemd_networkctl_result.stderr)
+            }
+        );
+
+        let status_output = String::from_utf8(systemd_networkctl_result.stdout)
+            .context(error::NetworkctlParsingSnafu {})?;
+
+        let networkd_status =
+            serde_json::from_str(&status_output).context(error::NetworkctlDeserializeSnafu {})?;
+
+        Ok(networkd_status)
+    }
+
+    // Fetches the IP Address for the primary interface. If there are no addresses, this is
+    // an error. If there is one, it should be returned. If there is more than one, then find
+    // the first IPv4 address and return it, if there are none, return the first IPv6 address.
+    pub(crate) fn primary_address(&self) -> Result<IpAddr> {
+        match self.addresses.len().cmp(&1) {
+            Ordering::Less => error::NoIpAddressSnafu {
+                interface: self.name.clone(),
+            }
+            .fail(),
+            Ordering::Equal => Ok(self.addresses[0]),
+            Ordering::Greater => {
```

Comment:
Is this just meant as an optimization? This comparison isn't really needed and you can just have the `Ordering::Greater` branch to remove some duplication.

Reply:
This is reasonable. I originally was focused on handling ensuring I handled all the cases but ultimately the logic would just work as you mention in the `Ordering::Greater` branch. It might be a marginal improvement to have this match but may not be worth the extra logic here. I can remove to make this more clear.